Changing datatype to Float64 for geoDistance (closes #58476)

geetptl opened the pull request:
### Changelog category:
- Improvement

<!---
I'm really unsure whether to put this in Improvement or bug-fix.
-->

### Changelog entry:
The geoDistance function now uses Float64 values for calculation to support high precision position values.

### Documentation entry for user-facing changes
There are shouldn't be any user-facing changes. (Open to correction here)

geetptl committed:
Changing datatype to Float64 for geoDistance

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=58847) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=58847) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=58847) it.</sub>

alexey-milovidov commented on src/Functions/greatCircleDistance.cpp:
```diff
@@ -295,13 +295,13 @@ class FunctionGeoDistance : public IFunction
             throw Exception(
                     ErrorCodes::ILLEGAL_COLUMN,
-                    "Illegal type {} of argument {} of function {}. Must be Float32.",
+                    "Illegal type {} of argument {} of function {}. Must be Float64.",
```
Can we allow both Float32 and Float64?

geetptl replied:
The function signature is of ColumnFloat64 though.

alexey-milovidov replied:
The logic should be as follows:
- if at least one argument is Float64, it uses the Float64 specialization;
- if all arguments are Float32, it uses the Float32 specialization.

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #58847 from geetptl/ClickHouse:master -->
*This is an automated comment for commit 38fa801f6743aa6fdf91ba6a602468b279b9f186 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58847/38fa801f6743aa6fdf91ba6a602468b279b9f186/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58847/38fa801f6743aa6fdf91ba6a602468b279b9f186/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl committed:
implicit conversion double to float error fix

geetptl committed:
Merge branch 'master' of github.com:geetptl/ClickHouse

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl committed:
Changing fabsf to std::abs

geetptl committed:
asinf fix, removing sqrf function

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl committed:
Float64 fix.

geetptl commented:
@alexey-milovidov I need help here on the failing tests. Since the precision is increased, the numbers do not match perfectly, hence the tests fail as following.

```
2024-01-16 10:59:54 --- /ClickHouse/tests/queries/0_stateless/01043_geo_distance.reference	2024-01-16 10:55:07.292471657 +0100
2024-01-16 10:59:54 +++ /ClickHouse/tests/queries/0_stateless/01043_geo_distance.stdout	2024-01-16 10:59:54.838310637 +0100
2024-01-16 10:59:54 @@ -1,8 +1,8 @@
2024-01-16 10:59:54 -111195.055
2024-01-16 10:59:54 -111195.055
2024-01-16 10:59:54 -110567.33
2024-01-16 10:59:54 -111699.25
2024-01-16 10:59:54 -10007554
2024-01-16 10:59:54 -10007555
2024-01-16 10:59:54 -10007554
2024-01-16 10:59:54 -10001780
2024-01-16 10:59:54 +111195.04881760638
2024-01-16 10:59:54 +111195.04881760638
2024-01-16 10:59:54 +110567.32686882635
2024-01-16 10:59:54 +111699.2516454354
2024-01-16 10:59:54 +10007554.393584572
2024-01-16 10:59:54 +10007554.393584574
2024-01-16 10:59:54 +10007554.393584572
2024-01-16 10:59:54 +10001780.1
2024-01-16 10:59:54 
2024-01-16 10:59:54 
2024-01-16 10:59:54 
2024-01-16 10:59:54 Database: test_zvx0zhn2
```

```
2024-01-16 10:59:58 --- /ClickHouse/tests/queries/0_stateless/01678_great_circle_angle.reference	2024-01-16 10:55:07.364473127 +0100
2024-01-16 10:59:58 +++ /ClickHouse/tests/queries/0_stateless/01678_great_circle_angle.stdout	2024-01-16 10:59:58.118376654 +0100
2024-01-16 10:59:58 @@ -1,5 +1,5 @@
2024-01-16 10:59:58  0.1224
2024-01-16 10:59:58  0.7071
2024-01-16 10:59:58  0.7135
2024-01-16 10:59:58 -10007555
2024-01-16 10:59:58 -10007554
2024-01-16 10:59:58 +10007554.3936
2024-01-16 10:59:58 +10007554.3936
2024-01-16 10:59:58 
2024-01-16 10:59:58 
2024-01-16 10:59:58 
2024-01-16 10:59:58 Database: test_2h3y3pjt
```

evillique commented on src/Functions/greatCircleDistance.cpp:
```diff
@@ -43,48 +43,43 @@ namespace
 
 constexpr double PI = std::numbers::pi_v<double>;
-constexpr float PI_F = std::numbers::pi_v<float>;
+constexpr double PI_F = std::numbers::pi_v<double>;
```
`PI_F` is supposed to mean PI float, so it would make more sense to just delete this and similar lines and replace all `PI_F` with `PI`

geetptl replied:
Gotcha, done.

evillique commented:
> @alexey-milovidov I need help here on the failing tests. Since the precision is increased, the numbers do not match perfectly, hence the tests fail as following.

After you update the `.reference` files for the failed tests they will not fail anymore :)

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl committed:
Removing PI_F

geetptl committed:
Test changes to reflect precision

geetptl commented:
> After you update the .reference files for the failed tests they will not fail anymore :)

Of course, I just wanted an approval before changing tests/results.

geetptl commented:
Are these failing integration tests and stateless tests unrelated? I fail to see how my changes would affect these tests.

geetptl committed:
Merge branch 'ClickHouse:master' into master

alexey-milovidov commented:
@geetptl, let's evaluate the performance.

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl commented:
@alexey-milovidov have a look at [this performance report](https://s3.amazonaws.com/clickhouse-test-reports/58847/9abfc8d29e376296307e7623317e55c860fbb327/performance_comparison_[4_4]/report.html#changes-in-performance)

alexey-milovidov commented:
I see - 11..12% difference.

Can we make two specializations of this function - Float32 and Float64? The less precise one will be used if at least one argument is Float32.

geetptl commented:
That would mean two separate implementations of same underlying functions using `float` and `double` data types, right?

alexey-milovidov commented:
Yes, but they can be unified in a template.

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl commented:
Got it.

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl commented:
Just doing some prototyping for understanding templates, and which would be a better approach from the following two?

First one is cleaner from an implementation and maintenance perspective, while the second one could be more optimised. What do you think?

```C++
#include <iostream>
#include <numbers>

enum class OtherParam
{
    o1,
    o2,
};

namespace ns
{
    constexpr double PI = std::numbers::pi_v<double>;

    template <OtherParam o, typename T>
    T increment_value_by_PI(T val)
    {
        std::cout << (o==OtherParam::o1) << std::endl;
        auto a = static_cast<T>(PI);
        return val + a;
    }
}

namespace ns2
{
    constexpr double PI = std::numbers::pi_v<double>;
    constexpr double PI_F = std::numbers::pi_v<float>;

    template <OtherParam o, typename T>
    T increment_value_by_PI(T val)
    {
        std::cout << (o==OtherParam::o1) << std::endl;
        if (typeid(val).name() == "f") {
            return val + PI_F;
        } else {
            return val + PI;
        }
    }
}

int main()
{
    double d = 1.5;
    float f = 1.5;

    std::cout << ns::increment_value_by_PI<OtherParam::o1>(d) << std::endl;
    std::cout << ns::increment_value_by_PI<OtherParam::o2>(f) << std::endl;

    std::cout << ns2::increment_value_by_PI<OtherParam::o1>(d) << std::endl;
    std::cout << ns2::increment_value_by_PI<OtherParam::o2>(f) << std::endl;
    return 0;
}
```

alexey-milovidov commented:
`if (typeid(val).name() == "f") {`

This won't work well.

geetptl committed:
Merge branch 'ClickHouse:master' into master

geetptl committed:
Template Changes
- Todo: Figure our where data types are deduced, so the correct template typename can be plugged

geetptl commented:
So far, I've made template related changes, and I'll see if there are any direct errors to be fixed.

I need your help for this though: Where in the greatCircleDistance function execution, data types are deduced from the column, so I can check whether double precision is required or not?

alexey-milovidov commented:
> Where in the greatCircleDistance function execution, data types are deduced from the column, so I can check whether double precision is required or not?

It should be in the `getReturnTypeImpl` function.

geetptl committed:
Merge branch 'ClickHouse:master' into master

alexey-milovidov commented on src/Functions/greatCircleDistance.cpp:
```diff
@@ -95,65 +89,70 @@ void geodistInit()
         return asin_sqrt_lut[i] + (asin_sqrt_lut[i + 1] - asin_sqrt_lut[i]) * (x - i);
     }
-    return asinf(sqrtf(x)); // distance over 17083 km, just compute exact
+    return asin(sqrt(x)); // distance over 17083 km, just compute exact
```
How does it dispatch between float/double sqrt and asin?

alexey-milovidov commented on src/Functions/greatCircleDistance.cpp:
```diff
@@ -95,65 +89,70 @@ void geodistInit()
+inline T geodistDegDiff(T f)
 {
-    f = fabsf(f);
+    f = abs(f);
```
Does it dispatch between float and double abs?

alexey-milovidov commented:
This looks abandoned. I will reimplement it: #61848
